# **Proyecto 3: Prueba de hipótesis - Déjame escuchar la música**

## **Contenido <a id='back'></a>**

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## **Descripción del proyecto**
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

## **Objetivo**
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


## **Etapas**
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

In [2]:
# Importar pandas

import pandas as pd

In [3]:
# Leer el archivo y almacenarlo en df

try:

    df = pd.read_csv('music_project_en.csv')

    print('Lectura exitosa')

except:

    print('Error de lectura')

Lectura exitosa


In [4]:
# Obtener las 10 primeras filas de la tabla df

print('Muestra las 10 primeras lineas de la tabla:\n')

print(df.head(10))

Muestra las 10 primeras lineas de la tabla:

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Mon

In [5]:
# Obtener la información general sobre nuestros datos

print('Información general de la tabla:\n')

print(df.info())

Información general de la tabla:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


**Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.**

Según la documentación:

- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

**Podemos ver tres problemas con el estilo en los encabezados de la tabla:**

1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`: **Corregir estilo de encabezados a snake_case para su estandarización.**




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

- En la información recabada encontramos solo datos de tipo objeto en las filas, lo que puede ocasionar problemas para realizar análisis cuantitativos de los mismos.

- No existe una columna que nos indique el género del usuario (H/M), en caso de que se quisiera profundizar en la comprobación del analisis por el género del usuario o usuaria.

- Existen datos ausentes en 3 columnas: 
  'Track': 1,343 
  'artist': 7,567 
  'genre' 1,198
  
- En la columna 'Day' solo considera 3 días de la semana: Wednesday, Friday, Monday.

- Existen 3,826 filas duplicadas.

- Se recomienda la busqueda de valores no estándar en los datos de la tabla.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [6]:
# Muestra los nombres de las columnas

print('Nombres de las columnas:\n')

print(df.columns)

Nombres de las columnas:

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

In [7]:
# Bucle en los encabezados poniendo todo en minúsculas


print('Nuevos encabezados de columnas:\n')

nuevos_encabezados = []

for col in df.columns:
    
    encabezado_minusculas = col.lower()
    
    nuevos_encabezados.append(encabezado_minusculas)

df.columns = nuevos_encabezados

print(df.columns)

Nuevos encabezados de columnas:

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [8]:
# Bucle en los encabezados eliminando los espacios

print('Eliminar espacios en nombres de columnas:\n')

nuevos_encabezados = []

for col in df.columns:
    
    encabezado_minusculas = col.lower()
    encabezado_stripped = encabezado_minusculas.strip()
    nuevos_encabezados.append(encabezado_stripped)

df.columns = nuevos_encabezados

print(df.columns)

Eliminar espacios en nombres de columnas:

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [9]:
# Cambiar el nombre de la columna "userid"

print('Aplicar estilo snake_case a los nombres de las columnas:\n')

nuevos_encabezados = []

for col in df.columns:
    
    encabezado_minusculas = col.lower()
    encabezado_stripped = encabezado_minusculas.strip()
    encabezado_snake = encabezado_stripped.replace('userid','user_id')
    nuevos_encabezados.append(encabezado_snake)

df.columns = nuevos_encabezados

print(df.columns)

Aplicar estilo snake_case a los nombres de las columnas:

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
# Comprobar el resultado: la lista de encabezados

print('Encabezados finales:\n')

nuevos_encabezados = []

for col in df.columns:
    
    encabezado_minusculas = col.lower()
    encabezado_stripped = encabezado_minusculas.strip()
    encabezado_snake = encabezado_stripped.replace('userid','user_id')
    nuevos_encabezados.append(encabezado_snake)

df.columns = nuevos_encabezados

print(df.columns)

Encabezados finales:

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [11]:
# Calcular el número de valores ausentes

print('Valores ausentes por columna:\n')

print(df.isna().sum())

Valores ausentes por columna:

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [12]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

col_a_remplazar = ['track', 'artist', 'genre']  # Columnas a remplazar

for col in col_a_remplazar:   #Iterar en las columnas
    
    df[col].fillna('unknown', inplace=True)  # Remplazar ausentes por 'unknown'

print('Valores columna "genre":\n')

print(df['genre'].sample(5))


Valores columna "genre":

2875             pop
25916    alternative
51998    alternative
55387     electronic
42682     electronic
Name: genre, dtype: object


C:\Users\the_o\AppData\Local\Temp\ipykernel_10520\610516491.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna('unknown', inplace=True)  # Remplazar ausentes por 'unknown'


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [13]:
# Contar valores ausentes

print('Valores ausentes despúes del remplazo:\n')

print(df.isna().sum())


Valores ausentes despúes del remplazo:

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [14]:
# Contar duplicados explícitos

print('Valores duplicados explícitos en la tabla:', df.duplicated().sum())


Valores duplicados explícitos en la tabla: 3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [15]:
# Eliminar duplicados explícitos

df = df.drop_duplicates().reset_index(drop=True)

print(df.head())


    user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [16]:
# Comprobar de nuevo si hay duplicados

print('Valores duplicados explícitos despúes de eliminación:', df.duplicated().sum())


Valores duplicados explícitos despúes de eliminación: 0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [17]:
# Inspeccionar los nombres de géneros únicos

print('Valores únicos en "genre":\n', df['genre'].value_counts().sort_index())
print()
print('Lista de géneros únicos del df:\n', df['genre'].unique())


Valores únicos en "genre":
 genre
acid            1
acoustic        5
action          4
adult          24
africa         16
             ... 
vocal          93
western        97
world        1947
worldbeat       2
ïîï             1
Name: count, Length: 269, dtype: int64

Lista de géneros únicos del df:
 ['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korea

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [18]:
# Función para reemplazar duplicados implícitos

def replace_wrong_genres (df, column, wrong_genres, correct_genre):
    
    """Remplaza los valores duplicados intrinsecos"""

    for wrong in wrong_genres:
        df[column] = df[column].replace(wrong_genres, correct_genre)
    
    return df 

wrong_genre = ['hip', 'hop', 'hip-hop']

correct = 'hiphop'

# Prueba de la función

df = replace_wrong_genres (df, 'genre', wrong_genre, correct)

print(df['genre'].sample(10))


37835         dance
19466           new
42248    electronic
45694        rusrap
22590          rock
47299           new
2818          dance
20737         dance
1635           jazz
51846          folk
Name: genre, dtype: object


Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [19]:
# Eliminar duplicados implícitos
    
def replace_wrong_genres (df, column, wrong_genres, correct_genre):
    
    """Remplaza los valores duplicados intrinsecos"""

    for wrong in wrong_genres:
        df[column] = df[column].replace(wrong_genres, correct_genre)
    
    return df 

wrong_genre = ['hip', 'hop', 'hip-hop']

correct = 'hiphop'

df = replace_wrong_genres (df, 'genre', wrong_genre, correct)

print('Muestra de la columna "genre":\n', df['genre'].sample(10))

Muestra de la columna "genre":
 47631       classical
27924            rock
59536            rock
53960    extrememetal
22138           indie
3527            world
27654       classical
58672            rock
53056             pop
43598           blues
Name: genre, dtype: object


Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [20]:
# Comprobación de duplicados implícitos

print('Valores únicos en "genre":\n', df['genre'].value_counts().sort_index())
print()
print('Lista de géneros únicos del df:')
print(df['genre'].unique())


Valores únicos en "genre":
 genre
acid            1
acoustic        5
action          4
adult          24
africa         16
             ... 
vocal          93
western        97
world        1947
worldbeat       2
ïîï             1
Name: count, Length: 266, dtype: int64

Lista de géneros únicos del df:
['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'kor

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

- Identificación de información duplicada explicita: utilizando el método .duplicated() fue posible identificar las filas que contenían información duplicada, es decir, que todas las celdas de un fila contaba con exactamente la misma información que otra considerada anteriormente.

- Identificación de duplicados implícitos: este paso requirió un análisis más profundo, especialmente en la columna ‘genre’, ya que se cuenta con una gran cantidad de géneros musicales en esta columna, por lo que fue importante determinar que géneros podría consolidarse en uno mas concreto, en este caso se realizó con el género ‘hiphop’.

- Realizar la limpieza de duplicados explícitos e implícitos permitirá tener un análisis mas fino de la información, ya que de lo contrario podríamos obtener desviaciones en el resultado final, diluyendo la asertividad de la información.


[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [21]:
# Contar las canciones reproducidas en cada ciudad

df_ciudad = df.groupby('city')['track'].count()

print(f'Canciones reproducidas por ciudad: \n {df_ciudad}')


Canciones reproducidas por ciudad: 
 city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


- En la ciudad de Springfield (42,741 canciones) se escuchan más del doble de canciones que las escuchadas en Shelbyville (18,512 canciones).

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [23]:
# Calcular las canciones reproducidas en cada uno de los tres días

df_dia = df.groupby('day')['track'].count()

print(f'Canciones reproducidas por día: \n {df_dia}')

Canciones reproducidas por día: 
 day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


- Usualmente los días viernes (21,840) y lunes (21,354) los usuarios pasan más tiempo reproduciendo música con una cantidad total similar de canciones reproducidas, el día miércoles (18,059) es el día con menor reproducciones de estos 3 días analizados.  

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [25]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks (df, day, city):    # Función filtro canciones por día y ciudad
    
    """Cuenta las canciones reproducidas por día y ciudad indicados"""


    df = df[df['day']== day]
    
    df = df[df['city'] == city]
    
    df = df['user_id'].count()
    
    return df

dia = 'Friday'

ciudad = 'Shelbyville'

canciones_dia_ciudad = number_tracks(df, dia, ciudad)

print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{ciudad}": \n {canciones_dia_ciudad}')

Canciones reproducidas el día "Friday" en la ciudad de "Shelbyville": 
 5895


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [36]:
# El número de canciones reproducidas en Springfield por día

def number_tracks (df, day, city):    # Función filtro canciones por día y ciudad

    """Cuenta las canciones reproducidas por día y ciudad indicados"""
    
    df = df[df['day']== day]
    
    df = df[df['city'] == city]
    
    df = df['user_id'].count()
    
    return df

dia = 'Monday'

ciudad = 'Springfield'


print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{ciudad}": {number_tracks(df, dia, ciudad)}')

dia = 'Wednesday'

print()

print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{
      ciudad}": {number_tracks(df, dia, ciudad)}')

dia = 'Friday'

print()

print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{
      ciudad}": {number_tracks(df, dia, ciudad)}')

Canciones reproducidas el día "Monday" en la ciudad de "Springfield": 15740

Canciones reproducidas el día "Wednesday" en la ciudad de "Springfield": 11056

Canciones reproducidas el día "Friday" en la ciudad de "Springfield": 15945


In [37]:
# El número de canciones reproducidas en Shelbyville por día

dia = 'Monday'

ciudad = 'Shelbyville'

print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{
      ciudad}": {number_tracks(df, dia, ciudad)}')

dia = 'Wednesday'

print()

print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{
      ciudad}": {number_tracks(df, dia, ciudad)}')

dia = 'Friday'

print()

print(f'Canciones reproducidas el día "{dia}" en la ciudad de "{
      ciudad}": {number_tracks(df, dia, ciudad)}')

Canciones reproducidas el día "Monday" en la ciudad de "Shelbyville": 5614

Canciones reproducidas el día "Wednesday" en la ciudad de "Shelbyville": 7003

Canciones reproducidas el día "Friday" en la ciudad de "Shelbyville": 5895


**Conclusiones**

Hipótesis:
1.	La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

Conclusiones: 

-	La hipótesis se puede aceptar parcialmente ya que la actividad si difiere de acuerdo con el día de la semana y ciudad:

    -Springfield mantiene el mayor número de usuarios los lunes (15,740) y viernes (15,945), una cantidad similar de usuarios en ambos días, y siendo su día más bajo el miércoles (11,056) con aproximadamente 5,000 escuchas menos.

    -Shelbyville su día pico máximo es el miércoles (7,003), los lunes (5,614) y viernes (5,895) son sus días bajos con un número similar de usuarios.

Adicional a lo anterior, la hipótesis se acepta parcialmente ya que sugerimos reforzar la misma realizando algunos análisis estadísticos adicionales como el tiempo promedio de escucha por día, ciudad y usuario ya que el número de reproducciones es subjetivo, una vez que puede ser que los usuarios reproducen una canción pero no terminan de escucharla, medir el tiempo promedio de reproducción nos podría indicar otra visión de la hipótesis, tal vez el tiempo promedio de escucha de música es menor o mayor en cierto día y ciudad. 

De igual forma se puede determinar si los géneros musicales más escuchados son similares o diferentes entre ambas ciudades.

Sugiero definir al menos 3 indicadores que soporten la decisión final acerca de la hipótesis:
1.	Número de reproducciones/usuarios por día-ciudad.
2.	Tiempo promedio de escucha por día-ciudad.
3.	Comparativo de top 10 géneros más escuchados por día-ciudad.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

Hipótesis:
1.	La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

Conclusiones: 

-	La hipótesis se puede aceptar parcialmente ya que la actividad si difiere de acuerdo con el día de la semana y ciudad:

    -Springfield mantiene el mayor número de usuarios los lunes (15,740) y viernes (15,945), una cantidad similar de usuarios en ambos días, y siendo su día más bajo el miércoles (11,056) con aproximadamente 5,000 escuchas menos.

    -Shelbyville su día pico máximo es el miércoles (7,003), los lunes (5,614) y viernes (5,895) son sus días bajos con un número similar de usuarios.

Adicional a lo anterior, la hipótesis se acepta parcialmente ya que sugerimos reforzar la misma realizando algunos análisis estadísticos adicionales como el tiempo promedio de escucha por día, ciudad y usuario ya que el número de reproducciones es subjetivo, una vez que puede ser que los usuarios reproducen una canción pero no terminan de escucharla, medir el tiempo promedio de reproducción nos podría indicar otra visión de la hipótesis, tal vez el tiempo promedio de escucha de música es menor o mayor en cierto día y ciudad. 

De igual forma se puede determinar si los géneros musicales más escuchados son similares o diferentes entre ambas ciudades.

Sugiero definir al menos 3 indicadores que soporten la decisión final acerca de la hipótesis:
1.	Número de reproducciones/usuarios por día-ciudad.
2.	Tiempo promedio de escucha por día-ciudad.
3.	Comparativo de top 10 géneros más escuchados por día-ciudad.


### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)